ext/ddtrace_profiling_native_extension/linux_tid_from_pthread.c

```diff
@@ -0,0 +1,209 @@
+
+  short setup_linux_tid_from_pthread_offset(void) {
+    int error;
+    struct tid_probe_info reference_threads_probe_info[REFERENCE_THREADS_COUNT];
+
+    // We use the current thread as the first reference thread
+    collect_tid_probe_info(&reference_threads_probe_info[0]);
+    if (!reference_threads_probe_info[0].buffer_read_success) return -1;
+
+    // ...and then we create a few more to act as extra references
+    for (int i = 1; i < REFERENCE_THREADS_COUNT; i++) {
+      pthread_t reference_thread;
+      error = pthread_create(&reference_thread, NULL, collect_tid_probe_info, &reference_threads_probe_info[i]);
+      if (error) return -1;
+      error = pthread_join(reference_thread, NULL);
+      if (error || !reference_threads_probe_info[i].buffer_read_success) return -1;
+    }
+
+    short candidate_tid_offset = 0;
+
+    // We try to find an offset that contains the expected tid for all reference threads. If we get to the end of the
+    // buffer before finding such an offset, the `thread_offset` will be -1 and we'll report back an error.
+    while (candidate_tid_offset >= 0) {
+      bool all_offsets_match = true;
+
+      for (int i = 0; i < REFERENCE_THREADS_COUNT; i++) {
+        short thread_offset = find_tid_offset_in_buffer(&reference_threads_probe_info[i], candidate_tid_offset);
+
+        if (thread_offset != candidate_tid_offset) {
+          all_offsets_match = false;
+          candidate_tid_offset = thread_offset;
+        }
+      }
+
+      if (all_offsets_match) break;
+    }
+
+    return candidate_tid_offset;
+  }
+
+  pid_t linux_tid_from(pthread_t thread, short offset) {
+    char buffer[STRUCT_PTHREAD_READ_SIZE];
+    read_struct_pthread(thread, buffer);
+    return get_tid_from_buffer(buffer, offset);
+  }
+
+  static void *collect_tid_probe_info(void *tid_probe_info) {
+    struct tid_probe_info *result = (struct tid_probe_info *) tid_probe_info;
+
+    result->tid = ddtrace_gettid();
+    result->buffer_read_success = read_struct_pthread(pthread_self(), result->buffer);
+
+    return NULL;
+  }
+
+  bool read_safely(void *read_from_ptr, void *read_into_buffer, short buffer_size) {
+    struct iovec read_into = {.iov_base = read_into_buffer, .iov_len = buffer_size};
+    struct iovec read_from = {.iov_base = read_from_ptr, .iov_len = buffer_size};
+
+    int unused_flags = 0;
+    int number_of_iovecs = 1;
+
+    short num_read = process_vm_readv(ddtrace_gettid(), &read_into, number_of_iovecs, &read_from, number_of_iovecs, unused_flags);
+
+    return num_read == buffer_size;
+  }
+
+  static bool read_struct_pthread(pthread_t thread, void *struct_pthread_buffer) {
+    return read_safely((void *) thread, struct_pthread_buffer, STRUCT_PTHREAD_READ_SIZE);
+  }
+
+  static short find_tid_offset_in_buffer(struct tid_probe_info *tid_probe_info, short start_position) {
+    if (!tid_probe_info->buffer_read_success) return -1;
+
+    short buffer_size = STRUCT_PTHREAD_READ_SIZE;
```

Comment:
Given we are reading 2048 bytes, is it possible that we read from a region of memory that belongs to a different data structure (beyond `struct pthread`)?

Reply:
It is! This is the cool thing about the `process_vm_readv` approach -- it really is reading how many bytes we ask, regardless of what's there, and returning an error if it couldn't.

So yeah, we may not read the whole structure, or we may grab extra stuff that isn't relevant. Worst case we don't find what we need (-1) or in very rare situations we may pick the wrong offset, but that seems quite rare seeing as we would need all 3 reference threads to agree on that wrong offset, and even then since this is only used for display the consequence is just wrong data in the UX.